### ISSUE TITLE
Rule: prevent global scope var/function pollution

### ISSUE THREAD

**Issue by josh** [eslint/eslint#4542]
I feel a little dumb asking this, but is there already a rule for preventing `var` and `function` global scope leaks in browsers?

The following patterns are considered problems:

``` js
function foo() {
}
var bar = 1;
```

The following patterns are considered not problems:

``` js
// using explicit window
window.foo = function() {
};
window.bar = 1;
```

``` js
// using iife
(function() {
  function foo() {
  }
  var bar = 1;
})();
```

Thanks!

**Comment by eslintbot** [comment]
@josh Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
What do you mean by "global scope leaks"? Do you mean you want a warning whenever a global variable is defined?

**Comment by josh** [comment]
> Do you mean you want a warning whenever a global variable is defined?

Yeah, without an explicit `window` or `self` assignment.

Its pretty common for developers to leak global helper functions if they are working outside an IIFE scope.

``` js
function onClick(el) {
}

$(".foo").on("click", onClick)
```

Suggest using an IIFE if they didn't mean for `onClick` to be global. And if they did, they should explicitly assign it as a `window` property.

**Comment by michaelficarra** [comment]
Only through top-level declarations in a script.

**Comment by josh** [comment]
> Only through top-level declarations in a script

Yep.

**Comment by josh** [comment]
@michaelficarra still used to the CoffeeScript world were a file based IIFE is implied 😄

**Comment by michaelficarra** [comment]
@josh Is GitHub finally dumping CoffeeScript?

**Comment by josh** [comment]
@michaelficarra haha, yeah. Thats the plan. Basically getting the new env setup with babel and eslint. @atom is making the same transition as well.

**Comment by ilyavolodin** [comment]
Sorry, I might be missing some context here. For browser script, what exactly is a difference between `window.foo = function() {...}` and `function foo() {...}` and why would you want to enforce first one? I would think that in most cases you would want the second one, since the script might be wrapped in the IIFE and in that case `foo` is not going to be polluting global scope.

**Comment by josh** [comment]
> For browser script, what exactly is a difference between window.foo = function() {...} and function foo() {...} and why would you want to enforce first one?

`window.foo` is explicitly setting a global. Most people don't realize `function foo() {...}` will be made a global if its defined at the root scope.

> since the script might be wrapped in the IIFE and in that case foo is not going to be polluting global scope.

You statically check to ensure thats the case.

``` js
// using iife, its okay
(function() {
  function foo() {
  }
  var bar = 1;
})();
```

**Comment by ilyavolodin** [comment]
Ahh.. Ok, so this is a stylistic rule request, to lower possible confusion. OK, got it.

**Comment by josh** [comment]
> Ahh.. Ok, so this is a stylistic rule request, to lower possible confusion. OK, got it.

Definitely.

**Comment by josh** [comment]
Theres also some interesting ES6 block scoping cases,

``` js
{
  // bad
  var a

  // ok, scoped to block, won't set a global
  let b
  const c
  function d() {}
}

// still bad
var x
let y
const z
```

**Comment by michaelficarra** [comment]
@ilyavolodin The goal is to force developers to explicitly distinguish between the two forms so you don't ever do one while expecting the behaviour of the other.

@josh It's about time. FYI, in the future, you can just use modules and then top-level declarations don't implicitly create properties on the global object.

**Comment by josh** [comment]
> FYI, in the future, you can just use modules and then top-level declarations don't implicitly create properties on the global object.

@michaelficarra yeah, I'm planning on keeping "scripts" and "modules" organized separately. Then only apply this rule in the "scripts" `.eslintrc`. None of our current code uses modules yet, but we'll make that transition in another step.

**Comment by michaelficarra** [comment]
@josh Actually, you said you were using babel. In that case, you should be using modules right now and this won't be a problem. Unless you need sloppy mode for something? But I doubt that.

**Comment by nzakas** [comment]
Seems like a good idea. @josh do you want to submit a PR? We are pretty shorthanded at the moment, so otherwise this may not get done.

**Comment by josh** [comment]
> @josh do you want to submit a PR?

Yeah, sure, I'll give it a shot!

**Comment by josh** [comment]
Any suggestions on the rule name? Seems like it should fall under the current "Variables" category. `no-global-vars`, `no-global-scope-vars`, `no-top-level-scope-vars`?

**Comment by mzgol** [comment]
@josh `no-implicit-globals`?

**Comment by nzakas** [comment]
:+1: no-implicit-globals

There are edge cases to be aware of when `ecmaFeatures.globalReturn` or `ecmaFeatures.modules` are true.

**Comment by michaelficarra** [comment]
What is the globalReturn edge case?

**Comment by nzakas** [comment]
Same as modules - the topmost scope is not global.

**Comment by michaelficarra** [comment]
I didn't know that that was implied by the globalReturn flag. I thought it just permitted top level returns in the parser.

**Comment by nzakas** [comment]
It's also used as a trigger to inject an extra scope to mimic Node.js scoping.
